File: section2/109.有序链表转换二叉搜索树.py
#
# @lc app=leetcode.cn id=109 lang=python3
#
# [109] 有序链表转换二叉搜索树
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __str__(self):
        return f'node(val={self.val})'

    @staticmethod
    def generate(nums: list):
        root = TreeNode(nums[0])
        nodes = [root]

        index = 1
        while index < len(nums):
            new_nodes = []
            for node in nodes:
                if node:
                    node.left = None if nums[index] is None else TreeNode(nums[index])
                    node.right = None if nums[index + 1] is None else TreeNode(nums[index + 1])

                new_nodes.append(node.left if node else None)
                new_nodes.append(node.right if node else None)
                index += 2

            nodes = new_nodes

        return root

    def show(self):
        stacks = [self]
        nodes = []
        while stacks:
            row, new_stacks = [], []
            for node in stacks:
                nodes.append(node.val if node else None)
                if node:
                    new_stacks.append(node.left)
                    new_stacks.append(node.right)

            stacks = new_stacks

        return nodes


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

    def __str__(self):
        return f'ListNode(val={self.val})'

    @staticmethod
    def generate(nums):
        head = ListNode(nums[0])
        node = head
        for num in nums[1:]:
            node.next = ListNode(num)
            node = node.next

        return head

class Solution:
    def find_mid(self, head, end):
        mid, fast = head, head
        while fast and fast != end and fast.next and fast.next != end:
            mid = mid.next
            fast = fast.next.next

        return mid

    def list_to_bst(self, head, end):
        if not head or head == end:
            return None

        mid = self.find_mid(head, end)
        node = TreeNode(mid.val)
        node.left = self.list_to_bst(head, mid)
        node.right = self.list_to_bst(mid.next, end)
        return node

    def sortedListToBST(self, head: ListNode) -> TreeNode:
        return self.list_to_bst(head, None)

    def test(self):
        print(self.sortedListToBST(ListNode.generate([-10, -3, 0, 5, 9])).show())


Solution().test()
# @lc code=end

File: section2/section/109.有序链表转换二叉搜索树.go
package section

import "fmt"

/*
 * @lc app=leetcode.cn id=109 lang=golang
 *
 * [109] 有序链表转换二叉搜索树
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func recursionSortedListToBST(head *ListNode, tail *ListNode) *TreeNode {
	if head == tail {
		return nil
	}

	slowNode, fastNode := head, head
	for fastNode != tail && fastNode.Next != tail {
		slowNode = slowNode.Next
		fastNode = fastNode.Next.Next
	}

	return &TreeNode{
		Val:   slowNode.Val,
		Left:  recursionSortedListToBST(head, slowNode),
		Right: recursionSortedListToBST(slowNode.Next, tail),
	}
}
func sortedListToBST(head *ListNode) *TreeNode {
	return recursionSortedListToBST(head, nil)
}

func Test109() {
	// 0,-3,9,-10,null,5
	fmt.Println(sortedListToBST(ToListNode([]int{-10, -3, 0, 5, 9})).ToValues())
}

// @lc code=end
